kernel/object_validate.c

```diff
@@ -0,0 +1,192 @@
+ * on the object already.
+ *
+ * @param object Address of kernel object
+ * @param thread Thread to grant access to the object
+ */
+void k_object_grant_access(void *object, struct k_thread *thread)
+{
+	struct _k_object *ko = _k_object_find(object);
+
+	if (!ko) {
+		printk("granting access to non-existent kernel object %p\n",
+		       object);
+		k_oops();
+	}
+	_k_object_grant_access(ko, thread);
+}
+
+
+/**
+ * Ensure a system object is a valid object of the expected type
+ *
+ * Searches for the object and ensures that it is indeed an object
+ * of the expected type, that the caller has the right permissions on it,
+ * and that the object has been initialized.
+ *
+ * This function is intended to be called on the kernel-side system
+ * call handlers to validate kernel object pointers passed in from
+ * userspace.
+ *
+ * @param obj Address of the kernel object
+ * @param otype Expected type of the kernel object
+ * @param init If true, this is for an init function and we will skip
+ *        permission and initialization state checks
+ * @return 0 If the object is valid
+ *         -EBADF if not a valid object of the specified type
+ *         -EPERM If the caller does not have permissions
+ *         -EINVAL Object is not intitialized
+ */
+int _k_object_validate(void *obj, enum k_objects otype, int init)
+{
+	struct _k_object *ko;
+
+	ko = _k_object_find(obj);
+
+	if (!ko || ko->type != otype) {
+#ifdef CONFIG_PRINTK
+		printk("%p is not a %s\n", obj, otype_to_str(otype));
+#endif
+		return -EBADF;
+	}
+
+	/* If true, this validation was called on behalf of an object
+	 * init function and we should skip remaining checks
+	 */
+	if (init) {
+		return 0;
+	}
+
+	if (!(ko->flags & K_OBJ_FLAG_INITIALIZED)) {
+		printk("%p used before initialization\n", obj);
+		return -EINVAL;
+	}
+
+	/* TODO: if the caller is in userspace, check that the proper bit
+	 * in ko->perms is set, otherwise return -EPERM
+	 */
+
+	return 0;
+}
+
+
+
+/**
+ * Lookup a kernel object and init its metadata if it exists
+ *
+ * Calling this on an object will make it usable from userspace.
+ * Intended to be called as the last statement in kernel object init
+ * functions.
+ *
+ * @param object Address of the kernel object
+ */
+void _k_object_init(void *object)
+{
+	struct _k_object *ko;
+
+	ko = _k_object_find(object);
+	if (!ko) {
+		/* Supervisor threads can ignore rules about kernel objects
+		 * and may declare them on stacks, etc. Such objects will never
+		 * be usable from userspace, but we shouldn't explode.
+		 */
+		return;
+	}
+
+	/* If this object was previously uninitialized, grant
+	 * permission for the calling thread
+	 */
+	if (!(ko->flags & K_OBJ_FLAG_INITIALIZED)) {
+		_k_object_grant_access(ko, _current);
+	} else {
+		/* TODO: If this object was already initialized, and the caller
+		 * was from userspace, and doesn't have permissions, k_oops()
+		 */
+	}
+
+	memset(ko->perms, 0, CONFIG_MAX_THREAD_BITS);
```

Comment:
In line 181, we just grant the access permission to the caller thread, but clear all permission bits here?

Reply:
yeah this isn't right. looks like i fixed it in some later work but I need to backport to this series.